src/execution_plan/execution_plan_clone.c

```diff
@@ -8,92 +8,69 @@
+	if(template->connected_components) {
+		array_clone_with_cb(clone->connected_components, template->connected_components, QueryGraph_Clone);
 	}
+
+	// Temporarily set the thread-local AST to be the one referenced by this ExecutionPlan segment.
+	QueryCtx_SetAST(clone->ast_segment);
+
 	return clone;
 }
 
-// Merge cloned execution plan segments, with respect to the plan type (union or not).
-static void _ExecutionPlan_MergeSegments(ExecutionPlan *plan) {
-	// No need to merge segments if there aren't any.
-	uint segment_count = array_len(plan->segments);
-	if(segment_count == 0) return;
-
-	if(plan->is_union) {
-		// Locate the join operation.
-		OpBase *join_op = ExecutionPlan_LocateOp(plan->root, OPType_JOIN);
-		assert(join_op);
-		// Each segment is a sub execution plan that needs to be joined.
-		for(int i = 0; i < segment_count; i++) {
-			ExecutionPlan *sub_plan = plan->segments[i];
-			ExecutionPlan_AddOp(join_op, sub_plan->root);
-		}
+static OpBase *_CloneOpTree(OpBase *template_parent, OpBase *template_current,
+							OpBase *clone_parent) {
+	const ExecutionPlan *plan_segment;
+	if(!template_parent || (template_current->plan != template_parent->plan)) {
+		/* If this is the first operation or it was built using a different ExecutionPlan
+		 * segment than its parent, clone the ExecutionPlan segment. */
+		plan_segment = _ClonePlanInternals(template_current->plan);
 	} else {
-		array_append(plan->segments, plan);
-		segment_count = array_len(plan->segments);
-		// Plan is not union, concatenate the segments.
-		OpBase *connecting_op;
-		// segments[0] is the first segment of the execution.
-		OpBase *prev_root = plan->segments[0]->root;
-		for(uint i = 1; i < segment_count; i++) {
-			ExecutionPlan *current_segment = plan->segments[i];
-			connecting_op = ExecutionPlan_LocateOpMatchingType(current_segment->root, PROJECT_OPS,
-															   PROJECT_OP_COUNT);
-			assert(connecting_op->childCount == 0);
-			ExecutionPlan_AddOp(connecting_op, prev_root);
-			prev_root = current_segment->root;
-		}
-		array_pop(plan->segments);
+		// This op was built as part of the same segment as its parent, don't change ExecutionPlans.
+		plan_segment = clone_parent->plan;
+	}
+
+	// Clone the current operation.
+	OpBase *clone_current = OpBase_Clone(plan_segment, template_current);
+
+	for(uint i = 0; i < template_current->childCount; i++) {
+		// Recursively visit and clone the op's children.
+		OpBase *child_op = _CloneOpTree(template_current, template_current->children[i], clone_current);
+		ExecutionPlan_AddOp(clone_current, child_op);
 	}
+
+	return clone_current;
+}
+
+static ExecutionPlan *_ExecutionPlan_Clone(const ExecutionPlan *template) {
+	OpBase *clone_root = _CloneOpTree(NULL, template->root, NULL);
+	// The "master" execution plan is the one constructed with the root op.
+	ExecutionPlan *clone = (ExecutionPlan *)clone_root->plan;
+	// Set the root op pointer.
+	clone->root = clone_root;
```

Comment:
isn't `clone_root->plan->root` is `clone_root` ?
in which case we can return `clone` ?

Reply:
This turns out to not be redundant, the `root` pointer is never touched by the cloning or the `New` op routines.